### PR TITLE
Stereo camera

### PR DESCRIPTION
**Pull request by BastiaanOlij**
These changes introduce a new camera mode to the camera object in Godot so it can be set as a uniform left/right/top/bottom frustum. The main reason for this is to allow asymmetrical frustums which is not possible with the normal camera settings but this is needed for left eye / right eye stereoscopic rendering. 

Note that the frustum is defined at unit distance NOT at the near plane. This allows the near and far plane to be adjusted without needing to change the frustum values themselves.
Also keep in mind that Godot will apply the aspect ratio of the viewport you are rendering to, you should not apply the aspect ratio to the frustum yourself (many textbook FOV -> frustum calculations will do this).

To help with this the PR also introduces a few functions on the camera that allow you to set the frustum based on the intraoculur distance (distance between the two eyes) and convergence distance (distance at which they eyes converge, basically anything in front of this will seem to pop out towards you, anything behind this distance will be seen as in depth).

I've got a little test project here:
https://github.com/BastiaanOlij/StereoCameraTest
This shows the stereoscopic rendering of two eyes to a split screen. I'll be adding on to this sample project in due time to highlight some of the differences between rendering for an HMD (if you want to use this for cardboard like mobile solutions) and stereo rendering on a 3D TV.

### CODE REVIEW

**Comment by reduz** [comment]
Nice, will inolement VR soon, so having this as reference to also implement
will be fantastic

**Comment by BastiaanOlij** [comment]
Reduz, I'm planning to start work on OpenVR tracking in the next week or two, just waiting on a mate who's Vive setup I can use. Theres a few issues like I've highlighted with getting the camera to work but this is a good starting point. The tracking should be pretty straight forward as the OpenVR/SteamVR API seems well though out.

Can anyone advise on the failure on Travis CL? It seems to have failed on the android build which is one setup I don't have to test on myself but one of the guys I've been testing this together with was doing Android builds without issue. There shouldn't be anything platform dependent in the code changes I've done here.

**Comment by akien-mga** [comment]
Android build errors (timeouts) are false positives, you can ignore them.

**Comment by BastiaanOlij** [comment]
I added a Frustum mode to the camera so I can support stereoscopic rendering on the Vive.  Won't know for sure if it is doing the right thing until I actually update the OpenVR code.

Note that the aspect ratio for the viewport is applied at rendering time so the frustum should be set assuming an aspect ratio of 1:1

I'm tempted to remove the stereo mode from the normal perspective camera settings and instead add a convenience method to configure the right frustum for each eye. That may end up giving cleaner code.

The implementation currently is still geared to Godot 2 and won't work on Godot 3 (hence the conflicts), I will get to that soon.

**Comment by BastiaanOlij** [comment]
Changed the solution around a bit. The camera in perspective mode has basically been brought back to how it was before these changes and I'm purely focussing on the added frustum approach. This has allowed me to return the visual server implementation almost back to how it was which should make updating this PR to Godot 3 a lot easier.

Setting the left and right eye projections now works through the set_perspective_for_eye method on the camera node. set_frustum lets you set any frustum. The frustum is sized at unity and without aspect ratio applied. Aspect ratio, near and far planes are all applied once the camera is used to render.

**Comment by BastiaanOlij** [comment]
The last check-in is just a small bug fix for something I found yesterday.

After a successful night last night working on my openvr project which uses these changes to do the correct stereoscopic rendering on a vive HMD I can conclude the changes finally work as designed. 

@akien-mga the changes here so far are still based on the 2.x master branch before this was merged in. Is it worth (and easy?) to move/duplicate the PR into the 2.x branch? There are a few people playing around with cardboard type things that might find this useful in 2.x.

I am planning to rebase this PR fairly soon and fix the conflicts with Godot 3.0 (that'll be a big one) now that I'm pretty certain I have what I need for openvr (which for the time being is still Godot 2)

**Comment by BastiaanOlij** [comment]
This latest check-in rebases my changes to Godot 3.0 but I haven't been able to test it because I'm getting a compile error unrelated to this work. One step closer though :)

File: core/math/camera_matrix.cpp [hunk 1]
```diff
@@ -94,8 +94,47 @@ void CameraMatrix::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_
...
+
+	// translate matrix by (modeltranslation, 0.0, 0.0)
+	CameraMatrix cm;
+	cm.set_identity();
+	cm.matrix[3][0] = modeltranslation;
+	*this = *this * cm;
+  
+}
+
+void CameraMatrix::set_orthogonal(real_t p_left, real_t p_right, real_t p_bottom, real_t p_top,  real_t p_znear, real_t p_zfar) {
```
Comment (BastiaanOlij) [inline core/math/camera_matrix.cpp]: Not sure why the diff messes up set_orthogonal, can anyone see what I changed that I shouldn't have changed?

File: servers/visual/visual_server_raster.cpp [hunk 2]
```diff
@@ -1641,7 +1641,6 @@ void VisualServerRaster::camera_set_perspective(RID p_camera,float p_fovy_degree
 	camera->fov=p_fovy_degrees;
 	camera->znear=p_z_near;
 	camera->zfar=p_z_far;
-
 }
 
 void VisualServerRaster::camera_set_orthogonal(RID p_camera,float p_size, float p_z_near, float p_z_far) {
```
Comment (BastiaanOlij) [inline servers/visual/visual_server_raster.cpp]: The merges in this file are really obsolete as all the code has moved into visual_server_scene but they got merged in from the Godot2 version of this change.

**Comment by BastiaanOlij** [comment]
Whoohoo, it compiles now:)

**Comment by BastiaanOlij** [comment]
I've added a simple frustum gizmo to the frustum camera mode. Not sure yet if I should extend it all the way to the far plane but it'll do for now.

It looks like left/right or up/down are reversed but thats my mistake I'll work on soon. I want to get some time behind the vive again to make sure which one is correctly reversed :)

**Comment by BastiaanOlij** [comment]
Frustum is now the right way around, looks like OpenVR has Y-axis in different direction then Godot.

**Comment by BastiaanOlij** [comment]
Just for future reference, I think it may have been in the original text

http://developer.download.nvidia.com/presentations/2009/SIGGRAPH/3DVision_Develop_Design_Play_in_3D_Stereo.pdf

This PDF is based on NVideas 3D Vision for 3DTV stereo projection but a lot of the information applies to HMDs as well. This gives some background info on why the frustum solution is needed.

**Comment by BastiaanOlij** [comment]
@reduz a little off topic but if my memory serves me correctly we had a talk about this on IRC awhile back. Just documenting it for future reference. 

At this moment the left camera and right camera are rendered to viewports individually. With the cameras nearly identical this could be optimised by doing most of the preparation once and then doing drawcalls for each object twice, once for the left eye, once for the right. Hardware capabilities will determine if this can be done object by object, or left eye first, then right eye, or in case of 3D vision capable hardware, duplicated on the hardware and rendering both viewports fully in parallel. 

The thing that is required is clipping planes that are able to clip out any object that isn't visible by both cameras. The near, far, top and bottom clipping planes are identical for both cameras, but the left and right clipping planes are different. However calculating a left and right clipping plane that encompasses both cameras is easy.

For the left clipping plane we take the position of the left eye and take the left position of the frustum at the far plane for the RIGHT eye.
For the right clipping plane we take the position of the right eye and take the right position of the frustum at the far plane of the LEFT eye. 

This creates a clipping volume that encompasses both eyes with as little dead space as possible and would allow us to create a list of object visible by both eyes and minimise the rendering.

File: core/math/frustum.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+#include "frustum.h"
+#include "math_funcs.h"
+
```
Comment (BastiaanOlij) [inline core/math/frustum.cpp]: @reduz , should this be changed to using real_t ? I noticed you made that change for most of the other classes as well.

Comment (bojidar-bg) [inline core/math/frustum.cpp]: Yeah, best if you change it, probably

Comment (BastiaanOlij) [inline core/math/frustum.cpp]: Done :)

**Comment by fluffrabbit** [comment]
Amazing! This kind of thing is exactly what Godot needs for VR in the long run. I'm going to take this for a test drive as soon as I can. I'm not touching the 3.0 branch for a couple months until it fully works, the 2.1 PR looks delicious, but the reason I'm commenting in this PR rather than the other is to respond to

>not rendering objects visible to the left eye but not visible to the right or visa versa

I think that would apply more to 3DTVs than HMDs. I don't want to imagine a virtual world in which objects which are potentially in front of other objects disappear or reappear from vision willy-nilly. Sounds too much like the old occlusion glitches with the Jak games and such. I know VR is a subjective thing, but the display technology should provide such a field of view and frustum as to model accurately what your eyes would see with the same field of view, like through binoculars or whatever depending on the FoV of the device.

**Comment by BastiaanOlij** [comment]
@fluffrabbit indeed, that is the thing that I constantly run into, there is a lot of contradicting info out there when you compare 3DTV requirements and HMD requirements. Its hard to know which one is right and which is wrong.

I also think there is a certain amount of "yes it should work like this but we can't just yet" with HMDs. For instance the convergence distance which for HMDs is best to leave at infinity. The asymmetric frustum is purely to deal with the fact that your eyes aren't in the middle of the screen for each eye.
But then you start looking at eye tracking like with the FOVE all of a sudden that gets put on its head. Suddenly you're adjusting the projection matrices each frame. 

As for the overlap in view frustums and proper clipping, that isn't an issue yet as right now each camera is rendered individually to its own render target. Each camera is clipped to each own view frustum. I have some ideas for adding stereoscopic render targets and doing single pass rendering where it does become an issue, and then we'll see what works best :)

**Comment by fluffrabbit** [comment]
Fascinating. Eye tracking could mean the camera for each eyeball points its own way, and the perspective distortion currently applied to turn the near and far clipping planes into curves may no longer apply on account of the eyeballs being curved (I don't know if I'm doing that math right) except to compensate for the shape of the headset. And don't holodecks work by projecting directly into the eyeballs?

What happens when you put a fancy HMD on a fish or a bird? Suddenly, it no longer makes sense to consider the eyes as sharing the same frustum because the shared frustum takes up almost 360 degrees. In fact, peripheral vision is such that if you either can see very very well or have a severe facial deformity (such as fetal alcohol syndrome or something) The combined FoV of your eyes is greater than 180 degrees, and what happens if you have a headset which covers that whole FoV? Then the shared frustum math does not work.

A shared frustum works for something like the Cardboard with parallel cameras, but if the eyes point in polar opposite directions it does not.

**Comment by BastiaanOlij** [comment]
Well we're still limited to the limits of the human eye and the eyes will still line up relatively well, its just that the convergence plane suddenly becomes more relevant as your eyes may be focussing on something close up and you can chose to do DOF effects better which can help combat motion sickness as things going towards your peripheral vision get blurred out. The shared frustum is more for discarding objects that are out of view to speed up rendering, ensure the engine only needs to do LOD decisions once per frame, and can better reuse dynamic shadow maps and such making sure you don't duplicate work just because the eyes are slightly apart. 
It does not have to be perfect as long as it doesn't discard too much, but a few false positives could easily be offset by not having to process everything twice per frame.

In the end the rendering of each eye still happens with their own frustums for clipping on the GPU and their own projection matrix. The only optimisation there is finding a way to render both eyes in parallel without introducing a lot of switching states between consecutive draw calls. Thats a discussion on its own for when we add a stereoscopic render buffer where you can have the choice between different approaches depending on hardware capabilities. 
As I understand it 3D vision for instance duplicates but caches all the drawcalls within the driver and batches a number of drawcalls for each eye to minimise this switching.
Most other engines currently render each eye in turn with just optimisations in preparing what to draw once.
I've been playing around with a technique to have a single double width render buffer and using an extra clipping plane so I can duplicate draw calls without doing any switching other then a single uniform that tells the shader which clipping plane and which projection matrix to use. In my test engine it hasn't yielded much improvements but it lacks a lot of the other optimisations to really test this. I'm hoping to find some time to try out the same technique in Godot to see if it would work and if it does, I'll submit it as a PR. 

That all said, you are on the money when you start looking at increasing the field of view. This is why you now have the headset (can't remember the name) that has 2 panels for each eye as its has side panels and the cameras to render those are completely different from the (near) forward facing panels. It gets even more complex if instead of doing the lens distortion as an after effect you actually want to properly render based on spherical projection. But I think that it will take a while before both GPU performance and HMDs catch up to a point where we can actually do that. 

But for the vive's, rift's, gearvr's, etc. of the world, that is all stuff we don't have to worry about.

**Comment by reduz** [comment]
As soon as I'm done with ports, we could probably try porting to vr

**Comment by BastiaanOlij** [comment]
@reduz well I got the basics working in Godot 2 (see my other PR) and have learned a lot in doing so. Find me on IRC one of these days so we can have a bit of a chat, there are various individual parts to this.

This PR is one that I think we can see as finished as its simply about setting up the camera in a way that makes more sense for stereoscopic rendering.

Then there is joining in the various VR SDKs, the openVR one I did for Godot 2 is a nice proof of concept but I think we need a different approach if we want Godot to be usable with various SDKs. I also had to cut some corners as I'm still new to Godot. There are also some things to learn by mistakes other engines have made. UE for instance is a real problem if you have both a Vive and a Rift as it has problems prioritising which SDK to use. 

Finally is optimising the rendering by introducing a stereoscopic render targets but even without that and just rendering each eye individually gets your very far.

**Comment by fluffrabbit** [comment]
I am not the person to be maintaining VR in a game engine. My aptitudes lie elsewhere. So even though it would be nice to make America first again or whatever it was I was planning with F-Droid, I have gone ahead and uploaded my mostly broken and embarassing pre-pre-alpha VR game to a repository so the people who actually know what they're doing can see what the state of VR with Godot is, this patch notwithstanding. Something better than homemade sensor fusion and two separate cameras would surely be an improvement.

https://github.com/fluffrabbit/SkyHammer Github didn't spray my TXZ files into the repo like I wanted it to so maybe that is for someone to do as a PR. Always happy add people as collaborators since like I said, I don't know what I'm doing, the knowing what one is doing is for other people. It's not like a job requirement to know because I'm my own boss, and at fluffrabbit LLC we only hire high school dropouts and felons. It keeps it fresh. Can't have any nerds around since they're terribly boring. Also can't have any dumbass graphic designers. The real artists are the ones who can work without computers. Oil paint was the original immersive experience.

I know I may have said something negative about a patch that added Android's orientation function as a sensor (Godot repo), but that is because the algorithms are invariably different between platforms, even if the sensors are the same. Until it's cross-platform, what's the point? It can't be Android-centric, and neither can it be Rift-centric. It has to work with all VR.

I hear the people working on OpenHMD have their hearts in the right places. It's all gonna be great. Remember WebVR-polyfill? Another example of greatness. The Web is shit, but if it has the power to bring all the different VR technologies under one roof, somebody is doing something right. OpenVR is controversial, OSVR is unnecessarily complex, and other VR libraries aim to unify all existing standards into a new standard like something out of an XKCD comic. There is a solution in it somewhere. Fun fact: OpenGL and OpenAL were both developed by greedy capitalist bastards. The thing with those is though they are libraries for market shares (Windows, Macs, Linux desktops). What is the market share with VR? Basically nobody uses it, and it's a real shame. Until there is a standalone HMD computer that is lightweight and runs something common like Android or Linux, and is available for the price of a netbook, I doubt the situation will change much. Hololens does not count, but just wait, as soon as Microsoft releases the consumer version, Google will rip them off with something better, and then the situation will change. Assuming they aren't too busy making laughably bad controllers and coming up with misguided product names.

Yes, VR is for gaming, and we need to accept that, but for that gamers should probably accept gaming platforms rather than deferring everything to the PC. There can be developer-friendly non-PC open gaming platforms. Vendor lock-in does not help the app ecosystem. Just look at the success of Android. Profitable? No. But it is the dominant mobile gaming platform. We need a dominant portable VR platform, or at least a VR platform that "just works" in the same way an Android phone does. Slipping a phone into a headset is awful. Plugging your head into a laptop is awful. Line-of-sight room-scale tracking systems are awful. We need a VR console that only does VR games. It's not a PC, it's not an accessory, it's a dedicated VR game machine that you keep under your bed for whenever you want to play a game, the thing you reach for before your GameBoy or PlayStation or laptop or TV, it's its own thing. That is the future.

Roughly 50% of the above probably makes sense. If I want to get fucked up that's my own business. You don't know what else is going on in my life.

**Comment by BastiaanOlij** [comment]
Just made a few small changes I had already applied to the 2.1 version of this PR including adding some documentation.

I also made a really boring video on the subject of using the new frustum camera mode to deal with stereo projection:
https://www.youtube.com/watch?v=tD06SpaeWOs

**Comment by fluffrabbit** [comment]
Very interesting, though I would describe it more as scientific than 
boring. I never gave IpD + screen size + screen distance much thought. I 
thought it was just for the distortion model. Didn't realize that the 
math you describe stops the eyes from skewing. I'm looking forward to 
the next one. I'm still too zoned out to test but it seems like this 
will get merged as so far it is the only stereo camera patch I've seen.

**Comment by fluffrabbit** [comment]
Also maybe non-flat VR screens are worth consideration. The optical 
gymnastics needed to bend light from a flat display surface to the full 
FoV > 180 degrees in the future seem preposterous. If typical human 
peripheral vision yields say 35 degrees behind the physical "near 
clipping plane" of your face, and the bridge of your nose gives you say 
25 degrees more to the left than the typical 90 degree FoV you get with 
certain Cardboard configurations, that's a lot of degrees, and assuming 
virtual cameras can only render up to 180 degrees per eye, you would 
need either 4 simultaneous cameras or pupil tracking. Also a VR headset 
designed for a bird couldn't possibly have a single display surface with 
the calculations you described. AFAIK there are already headsets on the 
market with split displays.

**Comment by BastiaanOlij** [comment]
Indeed, once you start talking about multiple displays or curved displays the game changes completely. Multiple displays in itself, if they are flat, aren't too big a deal, say that 4 display HMD, can't remember the name, you would have 4 cameras instead of 2 and instead of the projection plane being in the same plane, they will each will each have their own plane. It is all about mapping out where the screens are in relation to your eye and constructing the frustum accordingly. It all comes back to that first part of my video, the spectator not being in front of the display but viewing it from an angle.

Curved displays are an different ball game, its kinda in line with lens distortion if you want to do it properly. Lens distortion the "simple" way, what both the rift and vive do at the moment, involves rendering in a flat plane as per normal, and then stretching the image in post. Its actually very wasteful.

An alternative is to tesselate the 3D objects to a level where you can do a spherical projection in the vertex shader. Especially in combination with a tesselation shader to ensure you add enough detail in this can be a very effective way to render for a curved screen or to deal with lens distortion.

**Comment by BastiaanOlij** [comment]
@akien-mga, travis is complaining about clang formatting, probably because I hadn't  touched the 3.0 version of this PR for awhile. Not sure how to read what is actually wrong with it :)

**Comment by fluffrabbit** [comment]
FrustumCamera seems to be geared towards 3DTVs and HMDs whose combined 
FoV is less than 180 degrees, and the distortion is applied afterwards. 
It's well-designed but perhaps limited.

I'm imagining an optimal VR headset in a world without eye tracking. 
Each eyeball is encased in a para-hemispherical goggle such that the 
entire headset is made out of 2 spoon-shaped screens and 2 
unconventional lenses to cover the > 180 degree FoV for each eye. An 
eyeball's vertical FoV < 180 while its horizontal FoV > 180 due to the 
shape of the sockets, so naturally each screen receives a composite of 2 
cameras with a distortion shader applied. I guess that would be a total 
of 2 FrustumCameras for the headset.

It is theoretically possible to go the fragment shader route with Godot 
right now, but I'm not sure about tessellation in a vertex shader, that 
may have to wait for GLES3/Vulkan. I don't know much about shader 
efficiency. With say 100,000 polygons being tessellated to say 400,000 
I'm not sure whether or not that would be faster than a fragment shader.

**Comment by BastiaanOlij** [comment]
@fluffrabbit it is geared at devices that currently exist and techniques currently employed, and it is a technique that is as old as computer graphics are, I'm doing nothing new here. I just made it possible for Godot to go beyond the simpler symmetrical camera implementation so Godot can actually be used with todays HMDs.

Obviously you can start combining multiple render targets to create a larger FOV and piece it all together but you need a lot of GPU grunt, much more then is available on mobile, and even still challenging on a purpose build PC. Nvidia is tackling it at the moment by allow parallel rendering of multiple targets each with their own projection. They did a really good presentation on that a few months back.

However until there are devices that go beyond todays limitations and the industry decides on how to tackle and standardise rendering for those there is no point in building anything. If this becomes a big thing (or I should say when) I'm pretty sure we'll see graphic cards hit the market that have specific support for this and we'll see SDKs come out that have specific support for this. That is the time to start looking into adding support into Godot. Trying to pre-empt that is a waste of time really. 

All in all, this discussion is a completely separate one to have from what we're trying to accomplish here.

**Comment by fluffrabbit** [comment]
No, it's not a separate discussion, because HMDs with odd screen 
configurations exist today. While they may be corner cases and not 
currently "mainstream", more of their kind will probably hit the market 
within the next couple of years.

But do whatever you want. Like I said, I'm not in the right state of 
mind to test.

**Comment by fluffrabbit** [comment]
I tried to test but 3.0 wants a .godot file which is not in StereoCameraTest. I'm confused.

**Comment by BastiaanOlij** [comment]
@fluffrabbit , yes, that test project is still geared for Godot 2, so you'll need to load it up in Godot 2 and then export it for Godot 3 in order to test it, or just create a brand new scene.

**Comment by fluffrabbit** [comment]
I did that and got the following error when I tried to run it:

test_scene.gd:24

Parser Error: Name for local variable 'rotation' can't shadow class 
property of the same name.

**Comment by BastiaanOlij** [comment]
Yup, that's Godot 3.0 for you. It's really cool actually, in 3.0 properties are now directly accessible in GDScript but as there is a property "rotation" on any spatial object, you're no longer allowed to have a variable called rotation. So you spend some time renaming temporary variables after converting code from 2.0 to 3.0

**Comment by fluffrabbit** [comment]
My take on 3.0 is that the bright blue color scheme hurts my eyes, 
though a UI overhaul is long overdue to kill that shitty Windows 98 GUI 
Godot has had for so long. Anybody tried looking into Material Design? 
I'm all for minimalism if it makes things easier, plus it's not a bad 
idea to use colors to draw the eyes to where they need to go. We need a 
graphic designer or Steve Jobs type to fix things up. I'm getting 
migraines looking at the Godot interface.

I mean, Handbrake is a Mac-centric FLOSS and it's fairly 
straightforward, but better still is any given Android note-taking app 
on F-Droid. Sleek and efficient. Godot is just a shitton of buttons that 
makes me want to jump out my window, and I live on the 1st floor so it 
wouldn't even kill me, I'd just get a face full of glass. God, it just 
makes me sick.

Do whatever you gotta do. I'm assuming there's a version of this PR for 
2.1. Maybe you can make a branch of StereoCameraTest for Godot 3.0.

**Comment by BastiaanOlij** [comment]
Yes, PR #7778 is the Godot 2.1 PR.

Eventually I'll update StereoCameraTest for Godot 3 but since I'm doing most of this in 2 I'm not ready for that yet.. When 3 gets a stable build for iOS I'll give it some time. Its easy enough to setup.

**Comment by BastiaanOlij** [comment]
@akien-mga weird, even with the hooks in place it didn't complain about the last two small layout fixes. Just ran clang-format manually on these files, hopefully it'll be happy now...

**Comment by BastiaanOlij** [comment]
I've added one more function to the camera node:
`set_for_hmd(p_eye, p_intraocular_dist, p_display_width, p_display_to_lens, p_oversample, p_z_near, p_z_far)
`
This function sets the correct frustum for a given eye in an HMD environment where we're not getting the projection matrix from the SDK. It is based on:
- p_intraocular_dist: the intraocular distance, the distance between the two eyes (you will need to position the cameras accordingly)
- p_display_width: the width of the display for both eyes combined assuming side by side panels
- p_display_to_lens: the distance between the display and the lens
- p_oversample: the oversample or upscale. This increase the FOV based on the scale the render buffer is increased by, use the same scale when applying lens distortion.
- the near and far plane

**Comment by BastiaanOlij** [comment]
While i'm sure it will make some people sad the implementation for stereo scopic rendering for VR (#8567) is the way we've decided to go forward with this. Closing this PR.
